dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorPower.java

```diff
@@ -285,4 +296,336 @@ public List<Event> visitDat3mCAS(Dat3mCAS e) {
+
+        // Power does not have mo tags, thus we use null
+        Load load = newRMWLoadExclusive(dummyReg, address, null);
+        Store store = newRMWStoreExclusive(address, resultRegister, null, true);
+        Label label = newLabel("FakeDep");
+        Event fakeCtrlDep = newFakeCtrlDep(resultRegister, label);
+
+        Fence optionalMemoryBarrierBefore = mo.equals(Tag.Linux.MO_MB) ? Power.newSyncBarrier()
+                : mo.equals(Tag.Linux.MO_RELEASE) ? Power.newLwSyncBarrier() : null;
+        Fence optionalMemoryBarrierAfter = mo.equals(Tag.Linux.MO_MB) ? Power.newSyncBarrier()
+        		: mo.equals(Tag.Linux.MO_ACQUIRE)? Power.newISyncBarrier() : null;
+
+        
+        return eventSequence(
+                optionalMemoryBarrierBefore,
+                load,
+                localOp,
+                store,
+                fakeCtrlDep,
+                label,
+                optionalMemoryBarrierAfter
+        );
+	};
+
+	@Override
+	public List<Event> visitRMWFetchOp(RMWFetchOp e) {
+		Register resultRegister = e.getResultRegister();
+		IExpr value = (IExpr) e.getMemValue();
+		IExpr address = e.getAddress();
+		String mo = e.getMo();
+		
+        Register dummy = resultRegister;
+		if(resultRegister == value){
+            dummy = e.getThread().newRegister(resultRegister.getPrecision());
+        }
+
+		Load load = newRMWLoadExclusive(dummy, address, null);
+        Local optionalUpdateReg = dummy != resultRegister ? newLocal(resultRegister, dummy) : null;
+        Store store = newRMWStoreExclusive(address, new IExprBin(dummy, e.getOp(), value), null, true);
+        Label label = newLabel("FakeDep");
+        Event fakeCtrlDep = newFakeCtrlDep(resultRegister, label);
+
+        Fence optionalMemoryBarrierBefore = mo.equals(Tag.Linux.MO_MB) ? Power.newSyncBarrier()
+                : mo.equals(Tag.Linux.MO_RELEASE) ? Power.newLwSyncBarrier() : null;
+        Fence optionalMemoryBarrierAfter = mo.equals(Tag.Linux.MO_MB) ? Power.newSyncBarrier()
+        		: mo.equals(Tag.Linux.MO_ACQUIRE)? Power.newISyncBarrier() : null;
+
+        return eventSequence(
+        		optionalMemoryBarrierBefore,
+                load,
+                store,
+                optionalUpdateReg,
+                fakeCtrlDep,
+                label,
+                optionalMemoryBarrierAfter
+        );
+	}
+	
+	// Following
+	//		https://elixir.bootlin.com/linux/v5.18/source/arch/powerpc/include/asm/barrier.h
+	@Override
+	public List<Event> visitLoad(Load e) {
+		Register resultRegister = e.getResultRegister();
+		IExpr address = e.getAddress();
+		String mo = e.getMo();
+		
+        // Power does not have mo tags, thus we use null
+        Load load = newLoad(resultRegister, address, null);
+        Fence optionalMemoryBarrier = mo.equals(Tag.Linux.MO_ACQUIRE) ? Power.newLwSyncBarrier() : null;
+        
+        return eventSequence(
+                load,
+                optionalMemoryBarrier
+        );
+	}
```

Comment:
Before the code of this PR, `VisitorPower` would rely on the default implementation from `EventVisitor` for `Load`s, which is just `visitEvent()` which in turn just returns `Collections.singletonList(load);`. If the code does not contain LKMM events, then both implementation would match. However I agree this might  be dangerous if we are not cautious. The most reasonable solution I see is to introduce a new event `LKMMLoad` (similar to how we have `AtomicLoad`) to be sure that the code from above only applies to LKMM events.

Reply:
I don't think your code necessarily reduces to equivalent code if there are no LKMM events.
The difference is that the default implementation of `EventVisitor` will keep the load as is, while your implementation will create a new load. This new load may end up being different than the visited one. 
E.g., visiting an exclusive load will return a plain load. Also existing tags gets "deleted".

That being said, it is probably best to have `LKMMLoad` and friends as special events.